File: apps/mcp-endpoint-shield/vertex-ai-adk/typescript/callbacks.ts
/**
 * Akto Guardrails callbacks for Google Vertex AI ADK.
 *
 * Usage:
 *   import { aktoBeforeModelCallback, aktoAfterModelCallback } from "./callbacks";
 *
 *   const agent = new LlmAgent({
 *     model: "gemini-2.0-flash",
 *     name: "my_agent",
 *     instruction: "You are a helpful assistant.",
 *     beforeModelCallback: aktoBeforeModelCallback,
 *     afterModelCallback: aktoAfterModelCallback,
 *   });
 *
 * Environment Variables:
 *   DATA_INGESTION_URL   : Base URL for Akto's data ingestion service (required).
 *   SYNC_MODE            : "true" (default) to block before the LLM call;
 *                          "false" to validate/log asynchronously after the response.
 *   TIMEOUT              : HTTP request timeout in seconds (default: 5).
 *
 * Auto-detected (Vertex AI Agent Engine injects these at runtime):
 *   GOOGLE_CLOUD_PROJECT           : GCP project ID.
 *   GOOGLE_CLOUD_LOCATION          : Region (e.g. "us-central1").
 *   GOOGLE_CLOUD_AGENT_ENGINE_ID   : Reasoning Engine resource ID.
 */

import type { CallbackContext, LlmRequest, LlmResponse } from "@google/adk";
import type { Content, Part } from "@google/genai";

// ---------------------------------------------------------------------------
// Internal types
// ---------------------------------------------------------------------------

interface AktoSnapshot {
    model: string;
    messages: Record<string, unknown>[];
    blocked: boolean;
}

interface GuardrailsResult {
    allowed: boolean;
    reason: string;
}

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const DATA_INGESTION_URL: string | undefined = process.env.DATA_INGESTION_URL;
const SYNC_MODE: boolean =
    (process.env.SYNC_MODE || "true").toLowerCase() === "true";
const TIMEOUT: number = parseFloat(process.env.TIMEOUT || "5") * 1000; // ms
const AKTO_CONNECTOR_NAME = "vertex-ai-adk";
const HTTP_PROXY_PATH = "/api/http-proxy";

// State key used to pass request data from before_model_callback to after_model_callback.
const AKTO_SNAPSHOT_KEY = "__akto_request_snapshot";

console.info(
    `Akto Guardrails ADK callbacks initialized | sync_mode=${SYNC_MODE}`
);

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function getVertexEndpointInfo(): [string, string] {
    const project = process.env.GOOGLE_CLOUD_PROJECT;
    const location = process.env.GOOGLE_CLOUD_LOCATION;
    const engineId = process.env.GOOGLE_CLOUD_AGENT_ENGINE_ID;

    if (engineId && project && location) {
        // Agent Engine mode
        const host = `${location}-aiplatform.googleapis.com`;
        const path = `/v1/projects/${project}/locations/${location}/reasoningEngines/${engineId}:query`;
        return [host, path];
    }

    if (process.env.K_SERVICE) {
        // Cloud Run mode
        const serviceUrl = (process.env.CLOUD_RUN_SERVICE_URL || "")
            .replace(/^https?:\/\//, "")
            .replace(/\/+$/, "");
        return [serviceUrl || "run.googleapis.com", "/"];
    }

    // Local dev fallback
    return ["generativelanguage.googleapis.com", "/v1/chat/completions"];
}

function buildHttpProxyParams(
    guardrails: boolean,
    ingestData: boolean
): URLSearchParams {
    const params = new URLSearchParams();
    params.set("akto_connector", AKTO_CONNECTOR_NAME);
    if (guardrails) {
        params.set("guardrails", "true");
    }
    if (ingestData) {
        params.set("ingest_data", "true");
    }
    return params;
}

function parseGuardrailsResult(
    result: Record<string, unknown> | null
): GuardrailsResult {
    if (!result || typeof result !== "object") {
        return { allowed: true, reason: "" };
    }
    const data = result.data as Record<string, unknown> | undefined;
    if (!data || typeof data !== "object") {
        return { allowed: true, reason: "" };
    }
    const gr = data.guardrailsResult as Record<string, unknown> | undefined;
    if (!gr || typeof gr !== "object") {
        return { allowed: true, reason: "" };
    }
    const allowed = gr.Allowed !== undefined ? Boolean(gr.Allowed) : true;
    const reason = typeof gr.Reason === "string" ? gr.Reason : "";
    return { allowed, reason };
}

function contentsToMessages(
    contents: Content[] | undefined
): Record<string, unknown>[] {
    const messages: Record<string, unknown>[] = [];
    if (!contents) {
        return messages;
    }
    for (const content of contents) {
        if (!content) {
            continue;
        }
        const role = content.role || "user";
        const parts: Part[] = content.parts || [];
        const textParts = parts
            .filter((p) => p.text != null && p.text !== "")
            .map((p) => p.text!);
        messages.push({ role, content: textParts.join(" ") });
    }
    return messages;
}

function llmResponseToDict(
    llmResponse: LlmResponse | undefined
): Record<string, unknown> | null {
    if (!llmResponse || !llmResponse.content) {
        return null;
    }
    const content = llmResponse.content;
    const role = content.role || "model";
    const parts: Part[] = content.parts || [];
    const textParts = parts
        .filter((p) => p.text != null && p.text !== "")
        .map((p) => p.text!);
    return {
        choices: [{ message: { role, content: textParts.join(" ") } }],
    };
}

function makeBlockedResponse(reason: string): LlmResponse {
    const msg = reason
        ? `Blocked by Akto Guardrails: ${reason}`
        : "Blocked by Akto Guardrails";
    return {
        content: {
            role: "model",
            parts: [{ text: msg }],
        },
    };
}

function buildPayload(
    context: CallbackContext,
    model: string,
    messages: Record<string, unknown>[],
    responseBody: unknown,
    statusCode: number = 200
): Record<string, unknown> {
    const agentName = context.agentName || "unknown-agent";
    const invocationId = context.invocationId || "";
    const engineId = process.env.GOOGLE_CLOUD_AGENT_ENGINE_ID || process.env.K_SERVICE;
    const [host, path] = getVertexEndpointInfo();
    const timestamp = String(Date.now());
    const statusStr = String(statusCode);

    const modelBodyPayload = {
        model,
        messages,
    };

    const metadata = {
        call_type: "completion",
        model,
        agent_name: agentName,
        invocation_id: invocationId,
    };

    const tags = {
        "gen-ai": "Gen AI",
        "ai-agent": AKTO_CONNECTOR_NAME,
        "bot-name": engineId,
        source: "VERTEX_AI",
    };

    const requestPayload = JSON.stringify({
        body: JSON.stringify(modelBodyPayload),
    });
    const responsePayload =
        responseBody == null
            ? JSON.stringify({})
            : JSON.stringify({ body: JSON.stringify(responseBody) });

    return {
        path,
        requestHeaders: JSON.stringify({
            host,
            "content-type": "application/json",
        }),
        responseHeaders: JSON.stringify({
            "content-type": "application/json",
        }),
        method: "POST",
        requestPayload,
        responsePayload,
        ip: "0.0.0.0",
        destIp: "127.0.0.1",
        time: timestamp,
        statusCode: statusStr,
        type: "HTTP/1.1",
        status: statusStr,
        akto_account_id: "1000000",
        akto_vxlan_id: "0",
        is_pending: "false",
        source: "MIRRORING",
        direction: null,
        process_id: null,
        socket_id: null,
        daemonset_id: null,
        enabled_graph: null,
        tag: JSON.stringify(tags),
        metadata: JSON.stringify(metadata),
        contextSource: "AGENTIC",
    };
}

async function postHttpProxy(
    guardrails: boolean,
    ingestData: boolean,
    payload: Record<string, unknown>
): Promise<Response> {
    const params = buildHttpProxyParams(guardrails, ingestData);
    const endpoint = `${DATA_INGESTION_URL}${HTTP_PROXY_PATH}?${params.toString()}`;

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), TIMEOUT);

    try {
        return await fetch(endpoint, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(payload),
            signal: controller.signal,
        });
    } finally {
        clearTimeout(timeoutId);
    }
}

async function callGuardrailsValidation(
    context: CallbackContext,
    model: string,
    messages: Record<string, unknown>[]
): Promise<GuardrailsResult> {
    if (!DATA_INGESTION_URL) {
        return { allowed: true, reason: "" };
    }

    const payload = buildPayload(context, model, messages, null);
    try {
        const response = await postHttpProxy(true, false, payload);
        if (response.status !== 200) {
            console.info(
                `Guardrails validation returned HTTP ${response.status} (fail-open)`
            );
            return { allowed: true, reason: "" };
        }
        const result = (await response.json()) as Record<string, unknown>;
        return parseGuardrailsResult(result);
    } catch (e) {
        console.info(`Guardrails validation failed (fail-open): ${e}`);
        return { allowed: true, reason: "" };
    }
}

async function ingestResponsePayload(
    context: CallbackContext,
    model: string,
    messages: Record<string, unknown>[],
    responseBody: unknown,
    statusCode: number,
    logHttpError: boolean = false
): Promise<void> {
    if (!DATA_INGESTION_URL) {
        return;
    }

    const payload = buildPayload(context, model, messages, responseBody, statusCode);
    try {
        const response = await postHttpProxy(false, true, payload);
        if (logHttpError && response.status !== 200) {
            console.error(`Ingestion failed: HTTP ${response.status}`);
        }
    } catch (e) {
        console.error(`Ingestion failed: ${e}`);
    }
}

async function asyncValidateAndIngest(
    context: CallbackContext,
    model: string,
    messages: Record<string, unknown>[],
    responseDict: Record<string, unknown> | null
): Promise<void> {
    if (!DATA_INGESTION_URL) {
        return;
    }

    try {
        const payload = buildPayload(context, model, messages, responseDict);
        const response = await postHttpProxy(true, true, payload);
        if (response.status === 200) {
            const result = (await response.json()) as Record<string, unknown>;
            const { allowed, reason } = parseGuardrailsResult(result);
            if (!allowed) {
                console.info(
                    `Response flagged by guardrails (async mode, logged only): ${reason}`
                );
            }
        }
    } catch (e) {
        console.error(`Guardrails async validation error: ${e}`);
    }
}

// ---------------------------------------------------------------------------
// Public callback functions
// ---------------------------------------------------------------------------

/**
 * ADK before_model_callback — integrates Akto Guardrails pre-call validation.
 *
 * Behaviour depends on SYNC_MODE:
 *
 * - SYNC_MODE=true  (default):
 *     Sends the request to Akto's guardrails endpoint before forwarding it to
 *     the LLM. If the request is denied, a blocking LlmResponse is returned
 *     and the LLM is never called. The blocked interaction is also ingested
 *     for observability.
 *
 * - SYNC_MODE=false:
 *     Snapshots the request data in session state so that
 *     aktoAfterModelCallback can perform a combined validate-and-ingest
 *     call after the LLM responds. Returns undefined so the LLM call proceeds
 *     uninterrupted.
 *
 * Fail-open: any exception during validation allows the request through.
 */
export async function aktoBeforeModelCallback({
    context,
    request,
}: {
    context: CallbackContext;
    request: LlmRequest;
}): Promise<LlmResponse | undefined> {
    const model = request?.model || "";
    const messages = contentsToMessages(request?.contents);

    // Extract only the latest user message for guardrails validation.
    let latestUserMsg: Record<string, unknown>[] = [];
    for (const m of messages) {
        if (m.role === "user") {
            latestUserMsg = [m];
        }
    }
    if (latestUserMsg.length === 0 && messages.length > 0) {
        latestUserMsg = [messages[messages.length - 1]];
    }

    // Snapshot request data for aktoAfterModelCallback.
    context.state.set(AKTO_SNAPSHOT_KEY, {
        model,
        messages: latestUserMsg,
        blocked: false,
    } as AktoSnapshot);

    if (!SYNC_MODE) {
        return undefined;
    }

    try {
        const { allowed, reason } = await callGuardrailsValidation(
            context,
            model,
            latestUserMsg
        );
        if (!allowed) {
            // Ingest the blocked interaction before surfacing the error.
            await ingestResponsePayload(
                context,
                model,
                latestUserMsg,
                { "x-blocked-by": "Akto Proxy" },
                403
            );
            // Mark snapshot so after_model_callback skips duplicate ingestion.
            context.state.set(AKTO_SNAPSHOT_KEY, {
                model,
                messages: latestUserMsg,
                blocked: true,
            } as AktoSnapshot);
            console.info(`Request blocked by Akto Guardrails: ${reason}`);
            return makeBlockedResponse(reason);
        }
    } catch (e) {
        console.error(`Guardrails before-model error (fail-open): ${e}`);
    }

    return undefined;
}

/**
 * ADK after_model_callback — ingests or validates the completed interaction.
 *
 * Behaviour depends on SYNC_MODE:
 *
 * - SYNC_MODE=true  (default):
 *     Ingests the request + LLM response pair for observability.
 *
 * - SYNC_MODE=false:
 *     Sends a combined validate-and-ingest request. If the response is
 *     flagged by guardrails it is logged but NOT blocked (the LLM response
 *     is already being returned to the caller).
 *
 * Always returns undefined so the original LlmResponse is used unchanged.
 */
export async function aktoAfterModelCallback({
    context,
    response,
}: {
    context: CallbackContext;
    response: LlmResponse;
}): Promise<LlmResponse | undefined> {
    try {
        const snapshot = (context.state.get<AktoSnapshot>(AKTO_SNAPSHOT_KEY) || {}) as AktoSnapshot;
        const model = snapshot.model || "";
        const messages = snapshot.messages || [];

        // Skip if this request was already blocked and ingested in before_model_callback.
        if (snapshot.blocked) {
            return undefined;
        }

        const responseDict = llmResponseToDict(response);

        if (SYNC_MODE) {
            await ingestResponsePayload(
                context,
                model,
                messages,
                responseDict,
                200,
                true
            );
        } else {
            await asyncValidateAndIngest(context, model, messages, responseDict);
        }
    } catch (e) {
        console.error(`Guardrails after-model error: ${e}`);
    }

    return undefined;
}

File: apps/mcp-endpoint-shield/vertex-ai-adk/go/callbacks.go
// Akto Guardrails callbacks for Google Vertex AI ADK.
//
// Environment Variables:
//
//	DATA_INGESTION_URL          : Base URL for Akto's data ingestion service (required).
//	SYNC_MODE                   : "true" (default) to block before the LLM call;
//	                              "false" to validate/log asynchronously after the response.
//	TIMEOUT                     : HTTP request timeout in seconds (default: 5).
//
// Auto-detected (Vertex AI Agent Engine injects these at runtime):
//
//	GOOGLE_CLOUD_PROJECT        : GCP project ID.
//	GOOGLE_CLOUD_LOCATION       : Region (e.g. "us-central1").
//	GOOGLE_CLOUD_AGENT_ENGINE_ID: Reasoning Engine resource ID.
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"google.golang.org/adk/agent"
	"google.golang.org/adk/model"
	"google.golang.org/genai"
)

// ---------------------------------------------------------------------------
// Internal types
// ---------------------------------------------------------------------------

// aktoSnapshot is the internal state stored between before and after callbacks.
type aktoSnapshot struct {
	Model    string                   `json:"model"`
	Messages []map[string]interface{} `json:"messages"`
	Blocked  bool                     `json:"blocked"`
}

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

var (
	dataIngestionURL = os.Getenv("DATA_INGESTION_URL")
	syncMode         = strings.ToLower(getEnvDefault("SYNC_MODE", "true")) == "true"
	timeout          = parseTimeoutSeconds(getEnvDefault("TIMEOUT", "5"))
)

const (
	aktoConnectorName = "vertex-ai-adk"
	httpProxyPath     = "/api/http-proxy"
	aktoSnapshotKey   = "__akto_request_snapshot"
)

// ModelName should be set by the caller before using the callbacks.
// ADK v0.2.0 does not populate LLMRequest.Model — the model name is stored
// internally in the gemini model object and never passed to callbacks.
var ModelName string

// Shared HTTP client.
var httpClient = &http.Client{
	Timeout: timeout,
	Transport: &http.Transport{
		MaxIdleConns:        100,
		MaxIdleConnsPerHost: 20,
	},
}

func init() {
	log.Printf("[akto] callbacks initialized | data_ingestion_url=%q sync_mode=%v timeout=%v", dataIngestionURL, syncMode, timeout)
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

func getEnvDefault(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func parseTimeoutSeconds(s string) time.Duration {
	f, err := strconv.ParseFloat(s, 64)
	if err != nil {
		f = 5
	}
	return time.Duration(f * float64(time.Second))
}

func getVertexEndpointInfo() (host, path string) {
	project := os.Getenv("GOOGLE_CLOUD_PROJECT")
	location := os.Getenv("GOOGLE_CLOUD_LOCATION")
	engineID := os.Getenv("GOOGLE_CLOUD_AGENT_ENGINE_ID")

	if engineID != "" && project != "" && location != "" {
		// Agent Engine mode
		host = fmt.Sprintf("%s-aiplatform.googleapis.com", location)
		path = fmt.Sprintf("/v1/projects/%s/locations/%s/reasoningEngines/%s:query", project, location, engineID)
		return host, path
	}

	if kService := os.Getenv("K_SERVICE"); kService != "" {
		// Cloud Run mode
		serviceURL := os.Getenv("CLOUD_RUN_SERVICE_URL")
		serviceURL = strings.TrimPrefix(serviceURL, "https://")
		serviceURL = strings.TrimPrefix(serviceURL, "http://")
		serviceURL = strings.TrimRight(serviceURL, "/")
		if serviceURL == "" {
			serviceURL = "run.googleapis.com"
		}
		return serviceURL, "/"
	}

	// Local dev fallback
	return "generativelanguage.googleapis.com", "/v1/chat/completions"
}

func buildHTTPProxyParams(guardrails, ingestData bool) url.Values {
	params := url.Values{}
	params.Set("akto_connector", aktoConnectorName)
	if guardrails {
		params.Set("guardrails", "true")
	}
	if ingestData {
		params.Set("ingest_data", "true")
	}
	return params
}

func parseGuardrailsResult(result map[string]interface{}) (allowed bool, reason string) {
	if result == nil {
		return true, ""
	}
	data, ok := result["data"].(map[string]interface{})
	if !ok {
		return true, ""
	}
	gr, ok := data["guardrailsResult"].(map[string]interface{})
	if !ok || gr == nil {
		return true, ""
	}
	allowed = true
	if v, ok := gr["Allowed"].(bool); ok {
		allowed = v
	}
	if v, ok := gr["Reason"].(string); ok {
		reason = v
	}
	return allowed, reason
}

func contentsToMessages(contents []*genai.Content) []map[string]interface{} {
	log.Printf("[akto] contentsToMessages: converting %d content(s)", len(contents))
	var messages []map[string]interface{}
	for i, content := range contents {
		if content == nil {
			log.Printf("[akto]   content[%d]: nil, skipping", i)
			continue
		}
		role := string(content.Role)
		if role == "" {
			role = "user"
		}
		var textParts []string
		for _, p := range content.Parts {
			if p != nil && p.Text != "" {
				textParts = append(textParts, p.Text)
			}
		}
		text := strings.Join(textParts, " ")
		log.Printf("[akto]   content[%d]: role=%q text=%q", i, role, text)
		messages = append(messages, map[string]interface{}{
			"role":    role,
			"content": text,
		})
	}
	return messages
}

func llmResponseToDict(resp *model.LLMResponse) map[string]interface{} {
	if resp == nil || resp.Content == nil {
		return nil
	}
	role := string(resp.Content.Role)
	if role == "" {
		role = "model"
	}
	var textParts []string
	for _, p := range resp.Content.Parts {
		if p != nil && p.Text != "" {
			textParts = append(textParts, p.Text)
		}
	}
	return map[string]interface{}{
		"choices": []map[string]interface{}{
			{
				"message": map[string]interface{}{
					"role":    role,
					"content": strings.Join(textParts, " "),
				},
			},
		},
	}
}

func makeBlockedResponse(reason string) *model.LLMResponse {
	msg := "Blocked by Akto Guardrails"
	if reason != "" {
		msg = fmt.Sprintf("Blocked by Akto Guardrails: %s", reason)
	}
	return &model.LLMResponse{
		Content: &genai.Content{
			Role:  "model",
			Parts: []*genai.Part{{Text: msg}},
		},
	}
}

func mustJSON(v interface{}) string {
	b, err := json.Marshal(v)
	if err != nil {
		return "{}"
	}
	return string(b)
}

func buildPayload(
	ctx agent.CallbackContext,
	modelName string,
	messages []map[string]interface{},
	responseBody interface{},
	statusCode int,
) map[string]interface{} {
	agentName := ctx.AgentName()
	if agentName == "" {
		agentName = "unknown-agent"
	}
	invocationID := ctx.InvocationID()
	engineID := os.Getenv("GOOGLE_CLOUD_AGENT_ENGINE_ID")
	if engineID == "" {
		engineID = os.Getenv("K_SERVICE")
	}
	host, path := getVertexEndpointInfo()
	timestamp := strconv.FormatInt(time.Now().UnixMilli(), 10)
	statusStr := strconv.Itoa(statusCode)

	// Use a struct so JSON marshalling preserves field order (model before messages),
	// matching the Python payload exactly.
	type modelBody struct {
		Model    string                   `json:"model"`
		Messages []map[string]interface{} `json:"messages"`
	}
	requestPayload := mustJSON(map[string]interface{}{
		"body": mustJSON(modelBody{Model: modelName, Messages: messages}),
	})

	metadata := map[string]interface{}{
		"call_type":     "completion",
		"model":         modelName,
		"agent_name":    agentName,
		"invocation_id": invocationID,
	}

	// bot-name is nil when GOOGLE_CLOUD_AGENT_ENGINE_ID is not set, matching Python behaviour.
	var botName interface{}
	if engineID != "" {
		botName = engineID
	}
	tags := map[string]interface{}{
		"gen-ai":   "Gen AI",
		"ai-agent": aktoConnectorName,
		"bot-name": botName,
		"source":   "VERTEX_AI",
	}

	// Match Python: send "{}" string when there is no response body (before-model guardrails call).
	responsePayload := mustJSON(map[string]interface{}{})
	if responseBody != nil {
		responsePayload = mustJSON(map[string]interface{}{
			"body": mustJSON(responseBody),
		})
	}

	return map[string]interface{}{
		"path": path,
		"requestHeaders": mustJSON(map[string]interface{}{
			"host":         host,
			"content-type": "application/json",
		}),
		"responseHeaders": mustJSON(map[string]interface{}{
			"content-type": "application/json",
		}),
		"method":          "POST",
		"requestPayload":  requestPayload,
		"responsePayload": responsePayload,
		"ip":              "0.0.0.0",
		"destIp":          "127.0.0.1",
		"time":            timestamp,
		"statusCode":      statusStr,
		"type":            "HTTP/1.1",
		"status":          statusStr,
		"akto_account_id": "1000000",
		"akto_vxlan_id":   "0",
		"is_pending":      "false",
		"source":          "MIRRORING",
		"direction":       nil,
		"process_id":      nil,
		"socket_id":       nil,
		"daemonset_id":    nil,
		"enabled_graph":   nil,
		"tag":             mustJSON(tags),
		"metadata":        mustJSON(metadata),
		"contextSource":   "AGENTIC",
	}
}

func postHTTPProxy(guardrails, ingestData bool, payload map[string]interface{}) (*http.Response, error) {
	endpoint := dataIngestionURL + httpProxyPath
	params := buildHTTPProxyParams(guardrails, ingestData)

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("marshal payload: %w", err)
	}

	u, err := url.Parse(endpoint)
	if err != nil {
		return nil, fmt.Errorf("parse endpoint: %w", err)
	}
	u.RawQuery = params.Encode()

	log.Printf("[akto] POST %s (guardrails=%v ingest=%v) payload_size=%d bytes", u.String(), guardrails, ingestData, len(body))
	log.Printf("[akto]   payload_body: %s", string(body))

	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	start := time.Now()
	resp, doErr := httpClient.Do(req)
	elapsed := time.Since(start)
	if doErr != nil {
		log.Printf("[akto]   POST failed after %v: %v", elapsed, doErr)
		return nil, doErr
	}
	log.Printf("[akto]   POST response: status=%d latency=%v", resp.StatusCode, elapsed)
	return resp, nil
}

func callGuardrailsValidation(
	ctx agent.CallbackContext,
	modelName string,
	messages []map[string]interface{},
) (allowed bool, reason string) {
	log.Printf("[akto] callGuardrailsValidation: model=%q messages=%d", modelName, len(messages))
	if dataIngestionURL == "" {
		log.Printf("[akto]   DATA_INGESTION_URL is empty, skipping guardrails (allow)")
		return true, ""
	}

	payload := buildPayload(ctx, modelName, messages, nil, 200)
	resp, err := postHTTPProxy(true, false, payload)
	if err != nil {
		log.Printf("[akto]   guardrails validation failed (fail-open): %v", err)
		return true, ""
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		log.Printf("[akto]   guardrails returned HTTP %d (fail-open)", resp.StatusCode)
		return true, ""
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		log.Printf("[akto]   guardrails response parse error (fail-open): %v", err)
		return true, ""
	}

	log.Printf("[akto]   guardrails raw response: %s", mustJSON(result))
	allowed, reason = parseGuardrailsResult(result)
	log.Printf("[akto]   guardrails result: allowed=%v reason=%q", allowed, reason)
	return allowed, reason
}

func ingestResponsePayload(
	ctx agent.CallbackContext,
	modelName string,
	messages []map[string]interface{},
	responseBody interface{},
	statusCode int,
	logHTTPError bool,
) {
	log.Printf("[akto] ingestResponsePayload: model=%q messages=%d statusCode=%d", modelName, len(messages), statusCode)
	if dataIngestionURL == "" {
		log.Printf("[akto]   DATA_INGESTION_URL is empty, skipping ingestion")
		return
	}

	payload := buildPayload(ctx, modelName, messages, responseBody, statusCode)
	resp, err := postHTTPProxy(false, true, payload)
	if err != nil {
		log.Printf("[akto]   ingestion POST failed: %v", err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode == 200 {
		log.Printf("[akto]   ingestion succeeded")
	} else if logHTTPError {
		log.Printf("[akto]   ingestion failed: HTTP %d", resp.StatusCode)
	}
}

func asyncValidateAndIngest(
	ctx agent.CallbackContext,
	modelName string,
	messages []map[string]interface{},
	responseDict map[string]interface{},
) {
	log.Printf("[akto] asyncValidateAndIngest: model=%q messages=%d", modelName, len(messages))
	if dataIngestionURL == "" {
		log.Printf("[akto]   DATA_INGESTION_URL is empty, skipping async validate+ingest")
		return
	}

	payload := buildPayload(ctx, modelName, messages, responseDict, 200)
	resp, err := postHTTPProxy(true, true, payload)
	if err != nil {
		log.Printf("[akto]   async validate+ingest POST failed: %v", err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode == 200 {
		var result map[string]interface{}
		if err := json.NewDecoder(resp.Body).Decode(&result); err == nil {
			allowed, reason := parseGuardrailsResult(result)
			log.Printf("[akto]   async guardrails result: allowed=%v reason=%q", allowed, reason)
			if !allowed {
				log.Printf("[akto]   response flagged by guardrails (async mode, logged only): %s", reason)
			}
		} else {
			log.Printf("[akto]   async response parse error: %v", err)
		}
	} else {
		log.Printf("[akto]   async validate+ingest returned HTTP %d", resp.StatusCode)
	}
}

// ---------------------------------------------------------------------------
// Public callback functions
// ---------------------------------------------------------------------------

// AktoBeforeModelCallback is the ADK BeforeModelCallback that integrates Akto Guardrails
// pre-call validation.
//
// Signature matches llmagent.BeforeModelCallback:
//
//	func(agent.CallbackContext, *model.LLMRequest) (*model.LLMResponse, error)
//
// Register via:
//
//	llmagent.Config{
//	    BeforeModelCallbacks: []llmagent.BeforeModelCallback{callbacks.AktoBeforeModelCallback},
//	}
//
// Behaviour depends on SYNC_MODE:
//
//   - SYNC_MODE=true  (default):
//     Sends the request to Akto's guardrails endpoint before forwarding it to
//     the LLM. If the request is denied, a blocking LLMResponse is returned
//     and the LLM is never called. The blocked interaction is also ingested
//     for observability.
//
//   - SYNC_MODE=false:
//     Snapshots the request data in session state so that
//     AktoAfterModelCallback can perform a combined validate-and-ingest
//     call after the LLM responds. Returns nil so the LLM call proceeds
//     uninterrupted.
//
// Fail-open: any error during validation allows the request through.
func AktoBeforeModelCallback(
	ctx agent.CallbackContext,
	llmRequest *model.LLMRequest,
) (*model.LLMResponse, error) {
	log.Printf("[akto] ===== BeforeModelCallback START =====")
	log.Printf("[akto]   agent=%q sync_mode=%v", ctx.AgentName(), syncMode)

	modelName := ""
	var contents []*genai.Content
	if llmRequest != nil {
		modelName = llmRequest.Model
		contents = llmRequest.Contents
		log.Printf("[akto]   model=%q contents_count=%d", modelName, len(contents))
	} else {
		log.Printf("[akto]   llmRequest is nil")
	}
	if modelName == "" {
		modelName = ModelName
		log.Printf("[akto]   LLMRequest.Model was empty, using ModelName=%q", modelName)
	}
	messages := contentsToMessages(contents)
	log.Printf("[akto]   converted to %d message(s)", len(messages))

	// Extract only the latest user message for guardrails validation.
	var latestUserMsg []map[string]interface{}
	for _, m := range messages {
		if role, _ := m["role"].(string); role == "user" {
			latestUserMsg = []map[string]interface{}{m}
		}
	}
	if len(latestUserMsg) == 0 && len(messages) > 0 {
		latestUserMsg = messages[len(messages)-1:]
	}
	log.Printf("[akto]   latest user message for guardrails: %s", mustJSON(latestUserMsg))

	// Snapshot request data for AktoAfterModelCallback.
	_ = ctx.State().Set(aktoSnapshotKey, aktoSnapshot{
		Model:    modelName,
		Messages: latestUserMsg,
		Blocked:  false,
	})
	log.Printf("[akto]   snapshot saved (blocked=false)")

	if !syncMode {
		log.Printf("[akto]   async mode — skipping pre-LLM guardrails, returning nil")
		log.Printf("[akto] ===== BeforeModelCallback END (pass-through) =====")
		return nil, nil
	}

	allowed, reason := callGuardrailsValidation(ctx, modelName, latestUserMsg)
	if !allowed {
		log.Printf("[akto]   BLOCKED by guardrails — reason=%q, ingesting blocked interaction", reason)
		// Ingest the blocked interaction before surfacing the error.
		ingestResponsePayload(
			ctx, modelName, latestUserMsg,
			map[string]interface{}{"x-blocked-by": "Akto Proxy"},
			403, false,
		)
		// Mark snapshot so after_model_callback skips duplicate ingestion.
		_ = ctx.State().Set(aktoSnapshotKey, aktoSnapshot{
			Model:    modelName,
			Messages: latestUserMsg,
			Blocked:  true,
		})
		log.Printf("[akto] ===== BeforeModelCallback END (BLOCKED) =====")
		return makeBlockedResponse(reason), nil
	}

	log.Printf("[akto]   guardrails passed — forwarding to LLM")
	log.Printf("[akto] ===== BeforeModelCallback END (allowed) =====")
	return nil, nil
}

// AktoAfterModelCallback is the ADK AfterModelCallback that ingests or validates
// the completed interaction.
//
// Signature matches llmagent.AfterModelCallback:
//
//	func(agent.CallbackContext, *model.LLMResponse, error) (*model.LLMResponse, error)
//
// Register via:
//
//	llmagent.Config{
//	    AfterModelCallbacks: []llmagent.AfterModelCallback{callbacks.AktoAfterModelCallback},
//	}
//
// Behaviour depends on SYNC_MODE:
//
//   - SYNC_MODE=true  (default):
//     Ingests the request + LLM response pair for observability.
//
//   - SYNC_MODE=false:
//     Sends a combined validate-and-ingest request. If the response is
//     flagged by guardrails it is logged but NOT blocked (the LLM response
//     is already being returned to the caller).
//
// Always returns nil so the original LLMResponse is used unchanged.
func AktoAfterModelCallback(
	ctx agent.CallbackContext,
	llmResponse *model.LLMResponse,
	llmResponseError error,
) (*model.LLMResponse, error) {
	log.Printf("[akto] ===== AfterModelCallback START =====")
	log.Printf("[akto]   agent=%q sync_mode=%v", ctx.AgentName(), syncMode)

	if llmResponseError != nil {
		log.Printf("[akto]   LLM returned error: %v", llmResponseError)
	}

	snapshotRaw, err := ctx.State().Get(aktoSnapshotKey)
	if err != nil {
		log.Printf("[akto]   failed to get snapshot from state: %v — skipping", err)
		log.Printf("[akto] ===== AfterModelCallback END (no snapshot) =====")
		return nil, nil
	}
	snapshot, ok := snapshotRaw.(aktoSnapshot)
	if !ok {
		log.Printf("[akto]   snapshot type assertion failed (got %T) — skipping", snapshotRaw)
		log.Printf("[akto] ===== AfterModelCallback END (bad snapshot) =====")
		return nil, nil
	}
	log.Printf("[akto]   snapshot: model=%q messages=%d blocked=%v", snapshot.Model, len(snapshot.Messages), snapshot.Blocked)

	// Skip if this request was already blocked and ingested in before_model_callback.
	if snapshot.Blocked {
		log.Printf("[akto]   request was blocked in BeforeModelCallback — skipping duplicate ingestion")
		log.Printf("[akto] ===== AfterModelCallback END (already blocked) =====")
		return nil, nil
	}

	responseDict := llmResponseToDict(llmResponse)
	if responseDict != nil {
		log.Printf("[akto]   LLM response: %s", mustJSON(responseDict))
	} else {
		log.Printf("[akto]   LLM response is nil")
	}

	if syncMode {
		log.Printf("[akto]   sync mode — ingesting request+response pair")
		ingestResponsePayload(
			ctx, snapshot.Model, snapshot.Messages,
			responseDict, 200, true,
		)
	} else {
		log.Printf("[akto]   async mode — combined validate+ingest")
		asyncValidateAndIngest(
			ctx, snapshot.Model, snapshot.Messages, responseDict,
		)
	}

	log.Printf("[akto] ===== AfterModelCallback END =====")
	return nil, nil
}
